lib/collection/src/update_handler.rs

```diff
@@ -401,12 +458,33 @@ impl UpdateHandler {
                     {
                         continue;
                     }
+
+                    // Block optimization until CPU budget is available for it
+                    log::trace!(
+                        "Blocking optimization check, waiting for CPU budget to be available"
+                    );
+                    let desired_cpus = max_rayon_threads(max_indexing_threads);
+                    optimizer_cpu_budget.block_until_budget(desired_cpus);
+                    log::trace!("Continue with optimizations, new CPU budget available");
+
+                    // Determine optimization handle limit based on max handles we allow
+                    // Skip if we reached limit, an ongoing optimization that finishes will trigger this loop again
+                    let limit = max_handles.saturating_sub(optimization_handles.lock().await.len());
```

Comment:
Can you briefly explain how this `limit` interacts with the cpu budget?

Reply:
It is unrelated to CPU budget, it is a concrete limit of optimizations per shard that are allowed at a given time. It can be configured by the user. We have to keep this for backwards compatibility, otherwise I'd probably remove this extra limit. It no defaults to no limit at all (to then limit optimizations based on CPU budget).

I've added this comment:

```rust
// Not related to the CPU budget, but a different limit for the maximum number
// of concurrent concrete optimizations per shard as configured by the user in
// the Qdrant configuration.
```

This limit can be set with: `storage.optimizers.max_optimization_threads`

Reply:
Do we plan to deprecate this then, or it will still play along indefinitely?

Reply:
> Do we plan to deprecate this then

Didn't think about a schedule for deprecation yet. I guess it doesn't hurt to keep for a while.